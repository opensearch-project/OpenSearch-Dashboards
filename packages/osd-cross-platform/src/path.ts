/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import { execSync } from 'child_process';
import { basename, normalize, resolve } from 'path';
import {
  realpathSync as nativeRealpathSync,
  openSync,
  closeSync,
  existsSync,
  unlinkSync,
} from 'fs';

export const NAMESPACE_PREFIX = process.platform === 'win32' ? '\\\\?\\' : '';

/**
 * Get a standardized reference to a path
 * @param {string} path - the path to standardize
 * @param {boolean} [usePosix=true] - produce a posix reference
 * @param {boolean} [escapedBackslashes=true] - on Windows, double-backslash the reference
 * @param {boolean} [returnUNC=false] - produce an extended reference
 */
export const standardize = (
  path: string,
  usePosix: boolean = true,
  escapedBackslashes: boolean = true,
  returnUNC: boolean = false
) => {
  // Force os-dependant separators
  const normal = normalize(path);

  // Filter out in-browser executions as well as non-windows ones
  if (process?.platform !== 'win32') return normal;

  if (usePosix) return normal.replace(/\\/g, '/');
  else if (escapedBackslashes) return normal.replace(/\\/g, '\\\\');
  else if (returnUNC) return '\\\\?\\' + normal;
  return normal;
};

/**
 * Windows-only function that uses PowerShell to calculate the full path
 * @param {string} path
 * @private
 */
const getFullPathSync = (path: string) => {
  if (process.platform !== 'win32') return path;

  try {
    const fullName = execSync(`powershell "(Get-Item -LiteralPath '${path}').FullName"`, {
      encoding: 'utf8',
    })?.trim?.();

    // Make sure we got something back
    if (fullName?.length > 2) return fullName;
  } catch (ex) {
    // Do nothing
  }

  return path;
};

/**
 * Windows-only function that uses PowerShell and Com Object to calculate the 8.3 path
 * @param {string} path
 * @private
 */
const getShortPathSync = (path: string) => {
  if (process.platform !== 'win32') return path;

  try {
    const shortPath = execSync(
      `powershell "$FSO = New-Object -ComObject Scripting.FileSystemObject; $O = (Get-Item -LiteralPath '${path}'); if ($O.PSIsContainer) { $FSO.GetFolder($O.FullName).ShortPath } else { $FSO.GetFile($O.FullName).ShortPath }"`,
      {
        encoding: 'utf8',
      }
    )?.trim?.();

    // Make sure we got something back
    if (shortPath?.length > 2) return shortPath;
  } catch (ex) {
    // Do nothing
  }

  return path;
};

/**
 * Checks if Windows 8.3 short names are supported on the volume of the given path
 * @param {string} [path='.'] - the path to examine
 */
export const shortNamesSupportedSync = (path: string = '.') => {
  if (process.platform !== 'win32') return false;

  const testFileName = '.___osd-cross-platform-test.file';
  const file = resolve(path, testFileName);

  // Create a test file if it doesn't exist
  if (!existsSync(file)) closeSync(openSync(file, 'w'));

  // If the returned value's basename is not the same as the requested file name, it must be a short name
  const foundShortName = basename(getShortPathSync(file)) !== testFileName;

  // Cleanup
  unlinkSync(file);

  return foundShortName;
};

/**
 * @borrows shortNamesSupportedSync
 */
export const shortNameSupportedSync = shortNamesSupportedSync;

/**
 * Get the full pathname
 * @param {string} path - the path to resolve
 */
export const resolveToFullPathSync = (path: string) => getFullPathSync(resolve(path));

/**
 * @borrows resolveToFullPathSync
 */
export const resolveToFullNameSync = resolveToFullPathSync;

/**
 * Get the short pathname
 * @param {string} path - the path to resolve
 */
export const resolveToShortPathSync = (path: string) => getShortPathSync(resolve(path));

/**
 * @borrows resolveToShortPathSync
 */
export const resolveToShortNameSync = resolveToShortPathSync;

/**
 * Get the canonical pathname
 * @param {string} path - the path to resolve
 */
export const realPathSync = (path: string) => getFullPathSync(nativeRealpathSync(path, 'utf8'));

/**
 * @borrows realPathSync
 */
export const realpathSync = realPathSync;

/**
 * Get the canonical pathname
 * @param {string} path - the path to resolve
 */
export const realShortPathSync = (path: string) =>
  getShortPathSync(nativeRealpathSync(path, 'utf8'));

/**
 * @borrows realShortPathSync
 */
export const realshortpathSync = realShortPathSync;
