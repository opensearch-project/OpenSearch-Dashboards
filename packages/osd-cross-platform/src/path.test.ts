/*
 * Copyright OpenSearch Contributors
 * SPDX-License-Identifier: Apache-2.0
 */

import path from 'path';
import fs from 'fs';
import { access, rm, mkdir, writeFile, symlink } from 'fs/promises';

import {
  resolveToFullNameSync,
  resolveToFullPathSync,
  resolveToShortNameSync,
  resolveToShortPathSync,
  shortNamesSupportedSync,
  realPathSync,
  realShortPathSync,
  standardize,
} from './path';

const tmpTestFolder = './__test_artifacts__';
const longFolderName = '.long-folder-name';
const longFileName = '.long-file-name.txt';
const longSymlinkName = '.sym.link';
const shortFolderName = 'LONG-F~1';
const shortFileName = 'LONG-F~1.TXT';
const dummyWindowsPath = 'C:\\a\\b\\c';
const dummyWindowsPOSIXPath = 'C:/a/b/c';
const dummyPOSIXPath = '/a/b/c';

const onWindows = process.platform === 'win32' ? describe : xdescribe;
const onWindowsWithShortNames = shortNamesSupportedSync() ? describe : xdescribe;

// Save the real process.platform
const realPlatform = Object.getOwnPropertyDescriptor(process, 'platform')!;

describe('Cross Platform', () => {
  describe('path', () => {
    onWindows('on Windows', () => {
      onWindowsWithShortNames('when 8.3 is supported', () => {
        beforeAll(async () => {
          // Cleanup
          try {
            // If leftover artifacts were found, get rid of them
            await access(tmpTestFolder);
            await rm(tmpTestFolder, { recursive: true });
          } catch (ex) {
            // Do nothing; if `rm` failed, let the `mkdir` below throw the error
          }

          await mkdir(tmpTestFolder);
          await mkdir(path.resolve(tmpTestFolder, longFolderName));
          await writeFile(path.resolve(tmpTestFolder, longFolderName, longFileName), '');
          await symlink(
            path.resolve(tmpTestFolder, longFolderName),
            path.resolve(tmpTestFolder, longSymlinkName),
            'junction'
          );
        });

        afterAll(async () => {
          try {
            await rm(tmpTestFolder, { recursive: true });
          } catch (ex) {
            // Do nothing
          }
        });

        it('can synchronously extract full name of a folder', () => {
          const name = path.basename(
            resolveToFullPathSync(path.resolve(tmpTestFolder, shortFolderName))
          );
          expect(name).toBe(longFolderName);
        });

        it('can synchronously extract full name of a file', () => {
          const name = path.basename(
            resolveToFullNameSync(path.resolve(tmpTestFolder, shortFolderName, shortFileName))
          );
          expect(name).toBe(longFileName);
        });

        it('can synchronously extract short name of a folder', () => {
          const name = path.basename(
            resolveToShortPathSync(path.resolve(tmpTestFolder, longFolderName))
          );
          expect(name).toBe(shortFolderName);
        });

        it('can synchronously extract short name of a file', () => {
          const name = path.basename(
            resolveToShortNameSync(path.resolve(tmpTestFolder, longFolderName, longFileName))
          );
          expect(name).toBe(shortFileName);
        });

        it('can synchronously extract full name of a symbolic link', () => {
          const name = path.basename(realPathSync(path.resolve(tmpTestFolder, longSymlinkName)));
          expect(name).toBe(longFolderName);
        });

        it('can synchronously extract short name of a symbolic link', () => {
          const name = path.basename(
            realShortPathSync(path.resolve(tmpTestFolder, longSymlinkName))
          );
          expect(name).toBe(shortFolderName);
        });
      });
    });

    describe('on platforms other than Windows', () => {
      let mockPathNormalize: jest.SpyInstance<string, [p: string]>;
      let mockPathResolve: jest.SpyInstance<string, string[]>;
      let mockFSRealPathSync: jest.SpyInstance<string>;

      beforeAll(() => {
        Object.defineProperty(process, 'platform', {
          ...Object.getOwnPropertyDescriptor(process, 'property'),
          value: 'linux',
        });

        mockPathNormalize = jest.spyOn(path, 'normalize').mockReturnValue(dummyPOSIXPath);
        mockPathResolve = jest.spyOn(path, 'resolve').mockReturnValue(dummyPOSIXPath);
        mockFSRealPathSync = jest
          .spyOn(fs, 'realpathSync')
          .mockReturnValue(dummyPOSIXPath) as jest.SpyInstance<string>;
      });

      afterAll(() => {
        // Restore the real property value after each test
        Object.defineProperty(process, 'platform', realPlatform);
        mockPathNormalize.mockRestore();
        mockPathResolve.mockRestore();
        mockFSRealPathSync.mockRestore();
      });

      it('all short and full name methods return just the normalized paths', () => {
        expect(shortNamesSupportedSync()).toBe(false);
        expect(resolveToFullPathSync(dummyPOSIXPath)).toBe(dummyPOSIXPath);
        expect(resolveToShortPathSync(dummyPOSIXPath)).toBe(dummyPOSIXPath);
      });
    });

    describe('standardize', () => {
      describe('on Windows', () => {
        let mockPathNormalize: jest.SpyInstance<string, [p: string]>;

        beforeAll(() => {
          Object.defineProperty(process, 'platform', {
            ...Object.getOwnPropertyDescriptor(process, 'property'),
            value: 'win32',
          });

          mockPathNormalize = jest.spyOn(path, 'normalize').mockReturnValue(dummyWindowsPath);
        });

        afterAll(() => {
          // Restore the real property value after each test
          Object.defineProperty(process, 'platform', realPlatform);
          mockPathNormalize.mockRestore();
        });

        it('produces a path in native format', () => {
          expect(standardize(dummyWindowsPath, false, false)).toMatchSnapshot();
        });

        it('produces a path in native format even for POSIX input', () => {
          expect(standardize(dummyWindowsPOSIXPath, false, false)).toMatchSnapshot();
        });

        it('produces a path in native format with escaped backslashes', () => {
          expect(standardize(dummyWindowsPath, false, true)).toMatchSnapshot();
        });

        it('produces a path in POSIX format', () => {
          expect(standardize(dummyWindowsPath)).toMatchSnapshot();
        });
      });

      describe('on POSIX-compatible platforms', () => {
        let mockPathNormalize: jest.SpyInstance<string, [p: string]>;

        beforeAll(() => {
          Object.defineProperty(process, 'platform', {
            ...Object.getOwnPropertyDescriptor(process, 'property'),
            value: 'linux',
          });

          mockPathNormalize = jest.spyOn(path, 'normalize').mockReturnValue(dummyPOSIXPath);
        });

        afterAll(() => {
          // Restore the real property value after each test
          Object.defineProperty(process, 'platform', realPlatform);
          mockPathNormalize.mockRestore();
        });

        it('produces a path in POSIX format', () => {
          expect(standardize(dummyPOSIXPath)).toMatchSnapshot();
        });

        it('ignores additional parameters', () => {
          expect(standardize(dummyPOSIXPath, false, true)).toMatchSnapshot();
        });
      });
    });
  });
});
