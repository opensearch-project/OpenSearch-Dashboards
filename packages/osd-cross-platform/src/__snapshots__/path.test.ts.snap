// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cross Platform path standardize on POSIX-compatible platforms ignores additional parameters 1`] = `"/a/b/c"`;

exports[`Cross Platform path standardize on POSIX-compatible platforms produces a path in POSIX format 1`] = `"/a/b/c"`;

exports[`Cross Platform path standardize on Windows produces a path in POSIX format 1`] = `"C:/a/b/c"`;

exports[`Cross Platform path standardize on Windows produces a path in native format 1`] = `"C:\\\\a\\\\b\\\\c"`;

exports[`Cross Platform path standardize on Windows produces a path in native format even for POSIX input 1`] = `"C:\\\\a\\\\b\\\\c"`;

exports[`Cross Platform path standardize on Windows produces a path in native format with escaped backslashes 1`] = `"C:\\\\\\\\a\\\\\\\\b\\\\\\\\c"`;
